libraries/resource_providers/api200/volume_template_provider.rb

```diff
@@ -15,26 +15,33 @@ module API200
     class VolumeTemplateProvider < ResourceProvider
       # Loads the VolumeTemplate with all the external resources (if needed)
       def load_resource_with_associated_resources
-        raise "Unspecified property: 'storage_system'. Please set it before attempting this action." unless @new_resource.storage_system
-        raise "Unspecified property: 'storage_pool'. Please set it before attempting this action." unless @new_resource.storage_pool
         @item['provisioning']['capacity'] = @item['provisioning']['capacity'].to_s if @item['provisioning'] && @item['provisioning']['capacity']
-        load_storage_system
-        sp = resource_named(:StoragePool).find_by(@item.client, name: @new_resource.storage_pool, storageSystemUri: @item['storageSystemUri']).first
-        raise "Storage Pool '#{@new_resource.storage_pool}' not found for Storage System '#{@new_resource.storage_system}'" unless sp
-        @item['provisioning']['storagePoolUri'] = sp['uri']
+        @item.set_storage_system(load_storage_system)
+        @item['provisioning']['storagePoolUri'] = load_storage_pool['uri']
         @item.set_snapshot_pool(resource_named(:StoragePool).new(@item.client, name: @new_resource.snapshot_pool)) if @new_resource.snapshot_pool
       end
 
-      # Loads Storage System in the given VolumeTemplate resource.
+      # Loads the Storage Pool
+      # @raise [RuntimeError] if property 'storage_pool' is not set or Storage Pool resource not found
+      # @return [resource_named(:StoragePool)] The Storage Pool loaded
+      def load_storage_pool
```

Comment:
This whole method also feels like it could simply be replaced by a proper call to `load_resource`

Reply:
Nice! =D

Reply:
However, It is possible, in this case, use `load_resource ` only for StorageSystem but not for StoragePool, becase `load_resource ` call the `retrieve! ` method of Oneview resource, and this method will not return the expected when we have more than one `StoragePool ` with same name and with different `StorageSystem ` associated to.

This is the `retrieve!` method of `oneview-ruby-sdk`:
```ruby
    def retrieve!(header = self.class::DEFAULT_REQUEST_HEADER)
      retrieval_keys = self.class::UNIQUE_IDENTIFIERS.select { |k| !@data[k].nil? }
      raise IncompleteResource, "Must set resource #{self.class::UNIQUE_IDENTIFIERS.join(' or ')} before trying to retrieve!" if retrieval_keys.empty?
      retrieval_keys.each do |k|
        results = self.class.find_by(@client, { k => @data[k] }, self.class::BASE_URI, header)
        next if results.size != 1
        set_all(results[0].data)
        return true
      end
      false
    end
```

```ruby 
results = self.class.find_by(@client, { k => @data[k] }, self.class::BASE_URI, header) 
```
The line above just do the `find_by` using one key per time, but we need find the `StoragePool` by `name` and `storagePooolUri` at same time.

Reply:
@fgbulsoni Sorry, I have searched the `retrieve!` method into `StoragePool API200 ` and the `retrieve!` will work well in this case, because the StoragePool rewrote the method correctly. I will do the changes for StoragePool, too. =)

The `retrieve!` of `StoragePool API200`:
```ruby
      def retrieve!
        raise IncompleteResource, 'Must set resource name or uri before trying to retrieve!' unless @data['name'] || @data['uri']
        raise IncompleteResource, 'Must set resource storageSystemUri before trying to retrieve!' unless @data['storageSystemUri']
        results = self.class.find_by(@client, name: @data['name'], storageSystemUri: @data['storageSystemUri']) if @data['name']
        results = self.class.find_by(@client, uri:  @data['uri'], storageSystemUri: @data['storageSystemUri']) if @data['uri'] &&
          (!results || results.empty?)
        return false unless results.size == 1
        set_all(results[0].data)
        true
      end
```

Reply:
Sounds good, if StoragePool does not work greatly with .retrieve! the `load_resource` will indeed not be ideal.

Reply:
Awesome.